src/metabase/driver/postgres.clj

```diff
@@ -279,6 +285,49 @@
   (and (str/starts-with? database-type "\"")
        (str/ends-with? database-type "\"")))
 
+(defmethod sql.qp/->honeysql [:postgres :convert-timezone]
+  [driver [_ arg target-timezone source-timezone]]
+  (let [expr         (sql.qp/->honeysql driver arg)
+        timestamptz? (hx/is-of-type? expr "timestamptz")]
+    (when (and timestamptz? source-timezone)
+      (throw (ex-info (tru "`timestamp with time zone` columns shouldn''t have a `source timezone`")
+                      {:type            qp.error-type/invalid-query
+                       :target-timezone target-timezone
+                       :source-timezone source-timezone})))
+    (let [source-timezone (or source-timezone (qp.timezone/results-timezone-id))
+          ;; If a column doesn't have a timezone, calling `timezone(source-timezone, column)` will cast it
+          ;; to `timestamp with time zone` with time zone is the `source-timezone`.
+          ;; For `timestamp with time zone` columns, Postgres will shift the hours back to report-tz beforing returning it.
+          ;; So what we care is to just make sure the output of `convert-timezone` is a `timestammp with time zone`.
+          ;; And the actual conversion will be done in `format-rows` middleware.
+          expr (if-not timestamptz?
+                 (hsql/call :timezone source-timezone expr)
+                 expr)]
+      (hx/with-type-info expr
+        {::hx/convert-timezone {:source-timezone source-timezone
+                                :target-timezone target-timezone}
+         ::hx/database-type    "timestamptz"}))))
+
+(defn- to-timestamp-if-needed
+  "If the expr was converted to a timezone, convert to a timestamp so extract functions return date-part in the targeted timezone.
+
+  In Postgres, extract hour from a `timestamp with time zone` return the hour in `report-tz`.
+  I.e: select extract(hour, timestamp with time zone '2000/01/01 07:00:00+07:00') => 0 (assuming report-tz = UTC)
+
+  This is unintuive and our users will expect [:temporal-extract '2000/01/01 07:00:00+07:00' :hour] to returns 7 instead."
+  [expr]
+  (let [{:keys [target-timezone]} (hx/type-info->convert-timezone-info (hx/type-info expr))]
+    (if (and (hx/is-of-type? expr "timestamptz")
+             target-timezone)
+      (hx/with-database-type-info (hsql/call :timezone target-timezone expr) "timestamp")
```

Comment:
by calling `timezone` second time here, we'll:
- convert the input from `timestamp` to `timestamptz`
- The hours will be shifted to the `target-timezone`

So that means extracting an hour from the output of this will:
- returns the hour in `target-timezone` -- what we expected
- it'll automatically handle Daylight saving times.